File: app/callkit-java/src/main/java/com/example/callkit_java/CallKit.java
package com.example.callkit_java;

import android.content.Context;
import android.content.Intent;
import android.os.Build;

public class CallKit {

    private static CallKit instance = null;

    public Context context;
    public CallKitListener listener = null;

    private CallKit(Context context) {
        this.context = context;
    }

    public static CallKit getInstance(Context context) {
        if (instance == null) {
            instance = new CallKit(context);
        }
        return instance;
    }

    public void openIncomingCallScreen(String callerName, String number) {
        Intent intent = new Intent(context, CallkitService.class);
        intent.putExtra(CallkitService.CALLER_NAME, callerName);
        intent.putExtra(CallkitService.NUMBER, number);
        intent.setAction(HandleCallReceiver.ACTION_CREATE_CALL);
        context.startService(intent);
    }

    public void closeIncomingCallScreen() {
        context.stopService(new Intent(context, CallkitService.class));
    }

    public void setCallStatus(int status) {
        if (listener != null) {
            listener.onAcceptCall(status);
        }
    }

    public void setOnCallListener(CallKitListener listener) {
        this.listener = listener;
    }
}

File: app/callkit/src/main/java/com/example/callkit/CallKit.kt
package com.example.callkit

import SingletonHolder
import android.content.Context
import android.content.Intent
import android.os.Build

class CallKit private constructor(val context: Context) {
    var listener: CallKitListener? = null

    fun openIncomingCallScreen(caller: String? = "Hieu Nguyen", number: String? = "123456789") {
        val intent = Intent(context, CallkitService::class.java).apply {
            putExtra(CallkitService.CALLER_NAME, caller)
            putExtra(CallkitService.NUMBER, number)
            action = HandleCallReceiver.ACTION_CREATE_CALL
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(intent)
        } else {
            context.startService(intent)
        }
    }

    fun closeIncomingCallScreen() {
        context.stopService(Intent(context, CallkitService::class.java))
    }

    fun setCallStatus(status: Int) {
        listener?.onAcceptCall(status);
    }

    fun setOnCallListener(listener: CallKitListener?) {
        this.listener = listener
    }

    companion object : SingletonHolder<CallKit, Context>(::CallKit)
}